scripts/vscripts/chaos/effects/homingprojectiles.nut

```diff
@@ -0,0 +1,210 @@
+        FaceToward(new_target, self, projectile_speed)
+
+    return 0.02
+}
+
+::SelectVictim <- function(projectile)
+{
+    local target
+    local min_distance = 32000.0
+    for (local i = 1; i <= MaxClients(); i++)
+    {
+        local player = PlayerInstanceFromIndex(i)
+
+        if (player == null)
+            continue
+
+        local distance = (projectile.GetOrigin() - player.GetOrigin()).Length()
+
+        if (IsValidTarget(player, distance, min_distance, projectile))
+        {
+            target = player
+            min_distance = distance
+        }
+    }
+    return target
+}
+
+::IsLookingAt <- function(projectile, new_target)
+{
+    local target_origin = new_target.GetOrigin()
+    local projectile_owner = projectile.GetOwner()
+    local projectile_owner_pos = projectile_owner.EyePosition()
+
+    if (!projectile_owner.IsPlayer())
+         projectile_owner_pos = projectile_owner.GetCenter()
+
+    if (TraceLine(projectile_owner_pos, target_origin, projectile_owner) == 1)
+    {
+        local direction = (target_origin - projectile_owner.EyePosition())
+              direction.Norm()
+        local product   = projectile_owner.EyeAngles().Forward().Dot(direction)
+
+        if (product > 0.6)
+            return true
+            
+    }
+
+    return false
+}
+
+function AttachProjectileThinker(projectile)
+{
+    local projectile_speed = (projectile.GetAbsVelocity()).Norm()
+
+    projectile.ValidateScriptScope()
+    projectile.GetScriptScope().projectile_speed <- projectile_speed
+    AddThinkToEnt(projectile, "ProjectileThink")
+}
+
+::IsValidProjectile <- function(projectile)
+{
+    if (projectile.GetClassname() in validProjectiles)
+        return true
+    return false
+}
+
+::IsValidTarget <- function(victim, distance, min_distance, projectile)
+{
+    if (distance > min_distance || victim.GetHealth() == 0 || victim.GetTeam() == projectile.GetTeam() || NetProps.GetPropInt(victim, "m_lifeState") > 0)
+        return false
+    if (victim.IsPlayer() && (victim.IsInvulnerable() || victim.InCond(Constants.ETFCond.TF_COND_HALLOWEEN_GHOST_MODE) || victim.InCond(Constants.ETFCond.TF_COND_STEALTHED)))
+        return false
+
+    return true
+}
+
+::FaceToward <- function(new_target, projectile, projectile_speed)
+{
+    local desired_dir = new_target.EyePosition() - projectile.GetOrigin()
+          desired_dir.Norm()
+
+    local current_dir = projectile.GetForwardVector()
+    local new_dir = current_dir + (desired_dir - current_dir) * turn_rate
+          new_dir.Norm()
+
+    local move_ang = VectorAngles(new_dir)
+    local projectile_velocity = move_ang.Forward() * projectile_speed
+
+    projectile.SetAbsVelocity(projectile_velocity)
+    projectile.SetLocalAngles(move_ang)
+}
+
+// Only requiered for Detonator Flares and Bison as those persist even after hitting
+function Chaos_OnScriptHook_OnTakeDamage(params)
+{
+    if (params.inflictor.GetName() != "homing_projectile_XYZ2")

```

Comment:
Same as the comment from `ChaosEffect_Update`, why not keep an array of classnames that need this special handling and check it using `classnames.find(params.inflictor.GetClassname())`

Reply:
Performance impact and do not see a need as we only compare against 2 classnames, should this list grow however above that then we can use a table instead of an array as we used for checking validity too